src/cpp/session/modules/rmarkdown/SessionExecuteChunkOperation.hpp

```diff
@@ -0,0 +1,308 @@
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+#ifndef SESSION_MODULES_RMARKDOWN_SESSION_EXECUTE_CHUNK_OPERATIONR_HPP
+#define SESSION_MODULES_RMARKDOWN_SESSION_EXECUTE_CHUNK_OPERATIONR_HPP
+
+#include <boost/bind.hpp>
+#include <boost/enable_shared_from_this.hpp>
+#include <boost/noncopyable.hpp>
+#include <boost/shared_ptr.hpp>
+
+#include <core/FileSerializer.hpp>
+#include <core/text/CsvParser.hpp>
+#include <core/system/Process.hpp>
+#include <core/system/ShellUtils.hpp>
+#include <core/system/System.hpp>
+
+#include "NotebookOutput.hpp"
+#include "NotebookExec.hpp"
+#include "SessionRmdNotebook.hpp"
+
+namespace rstudio {
+namespace session {
+namespace modules {
+namespace rmarkdown {
+namespace notebook {
+
+core::shell_utils::ShellCommand shellCommandForEngine(
+      const std::string& engine,
+      const core::FilePath& scriptPath)
+{
+   core::shell_utils::ShellCommand command(engine);
+   
+   if (engine == "Rscript")
+      command << "-f";
+   
+   command << scriptPath;
+   
+   return command;
+}
+
+class ExecuteChunkOperation : boost::noncopyable,
+                              public boost::enable_shared_from_this<ExecuteChunkOperation>
+{
+   typedef core::shell_utils::ShellCommand ShellCommand;
+   typedef core::system::ProcessCallbacks ProcessCallbacks;
+   typedef core::system::ProcessOperations ProcessOperations;
+   
+public:
+   static boost::shared_ptr<ExecuteChunkOperation> create(const std::string& docId,
+                                                          const std::string& chunkId,
+                                                          const ShellCommand& command)
+   {
+      boost::shared_ptr<ExecuteChunkOperation> pProcess =
+            boost::shared_ptr<ExecuteChunkOperation>(new ExecuteChunkOperation(
+                                                        docId,
+                                                        chunkId,
+                                                        command));
+      pProcess->registerProcess();
+      return pProcess;
+   }
+   
+private:
+   
+   ExecuteChunkOperation(const std::string& docId,
+                         const std::string& chunkId,
+                         const ShellCommand& command)
+      : isRunning_(false),
+        terminationRequested_(false),
+        docId_(docId),
+        chunkId_(chunkId),
+        command_(command)
+   {
+      using namespace core;
+      Error error = Success();
+      
+      // ensure staging directory
+      FilePath stagingPath = chunkOutputPath(
```

Comment:
This shouldn't be necessary (did you see errors without it?). We use the staging path only as a container to store plots and HTML widgets as they are emitted.